codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -250,79 +251,168 @@ private void writePendingBytes() throws Http2Exception {
-                    // If this subtree still wants to send then re-insert into children list and re-consider for next
-                    // iteration. This is needed because we don't yet know if all the peers will be able to use
-                    // all of their "fair share" of the connection window, and if they don't use it then we should
-                    // divide their unused shared up for the peers who still want to send.
-                    if (state.streamableBytesForTree() > 0) {
-                        children[nextTail++] = child;
-                        nextTotalWeight += weight;
+
+            return nextConnectionWindow > 0;
+        }
+
+        void feedHungryChildren() throws Http2Exception {
+            if (stillHungry == null) {
+                // There are no hungry children to feed.
+                return;
+            }
+
+            totalWeight = nextTotalWeight;
+            connectionWindow = nextConnectionWindow;
+
+            // Loop until there are not bytes left to stream or the connection window has collapsed.
+            for (int tail = nextTail; tail > 0 && connectionWindow > 0;) {
+                nextTotalWeight = 0;
+                nextTail = 0;
+
+                // Iterate over the children that are currently still hungry.
+                for (int head = 0; head < tail && nextConnectionWindow > 0; ++head) {
+                    Http2Stream child = stillHungry[head];
+                    if (!visit(child)) {
+                        // The connection window has collapsed, break out of the loop.
+                        break;
                     }
                 }
+                connectionWindow = nextConnectionWindow;
+                totalWeight = nextTotalWeight;
+                tail = nextTail;
+            }
+        }
 
-                if (bytesForTree > 0) {
-                    int childBytesAllocated = allocateBytesForTree(child, bytesForTree);
-                    bytesAllocated += childBytesAllocated;
-                    nextConnectionWindow -= childBytesAllocated;
-                }
+        /**
+         * Indicates that the given child is still hungry (i.e. still has streamable bytes that can
+         * fit within the current connection window).
+         */
+        void stillHungry(Http2Stream child) {
+            allocateSpace(nextTail);
+            stillHungry[nextTail++] = child;
+        }
+
+        /**
+         * Ensures that the {@link #stillHungry} array is properly sized to hold the given index.
+         */
+        void allocateSpace(int index) {
+            if (stillHungry == null) {
+                // Initial size is 1/4 the number of children.
+                int minSize = max(index + 1, 4);
+                int desiredSize = maxSize / 4;
+                int initialSize = min(max(minSize, desiredSize), maxSize);
+                stillHungry = new Http2Stream[initialSize];
+            } else if (index >= stillHungry.length) {
```

Comment:
This conditional may be a bit misleading.  If `nextTail > childArray.length` then the following resize operation may not be sufficient.  For example `nextTail = 20, childArray.length = 4, children.size() = 40`...the array would only be sized to `8` and we would have an array out of bounds....not that this can realistically happen but the point is I think we can check for `nextTail == childArray.length` because that is what this class is really designed for.